fixes `CupertinoFullscreenDialogTransition` leaks [prod-leak-fix]

Dimilkalathiya opened the pull request:
Part of #141198 

Fixes`CupertinoFullscreenDialogTransition` widget leaks

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

Dimilkalathiya committed:
fixes `CupertinoFullscreenDialogTransition` leaks

Dimilkalathiya commented:
cc: @polina-c this one reduces leaks by 330 :)

Dimilkalathiya committed:
add `linearTransition` condition

polina-c commented:
Looks good. Can you opt-in some test, to make bot stop complaining about test coverage?

ValentinVignal commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -453,44 +453,95 @@ class CupertinoPageTransition extends StatelessWidget {
+  }
+
+  @override
+  void didUpdateWidget(covariant CupertinoFullscreenDialogTransition oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.primaryRouteAnimation != widget.primaryRouteAnimation
+    || oldWidget.secondaryRouteAnimation != widget.secondaryRouteAnimation
+    || oldWidget.child != widget.child || oldWidget.linearTransition != widget.linearTransition) {
```
I think you need to remove `oldWidget.child != widget.child`, this is not used to create the animation

polina-c replied:
@Dimilkalathiya , can you respond this message? Why we recalculate animations if child is changed?

Dimilkalathiya replied:
@polina-c  if `oldWidget.child != widget.child` not there it changes behaviour of animation offset i am not entirely sure is to why this is happening first i thought it might be performance improvement, but offsets are quite far to consider just that, so i decided to let it be there since previous implementation was also rebuilding widgets, would that be a problem?

polina-c replied:
Ok, thanks, let it stay there.

ValentinVignal replied:
Isn't it the same as what is happening in my PR ?

ValentinVignal commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -453,44 +453,95 @@ class CupertinoPageTransition extends StatelessWidget {
+         parent: widget.primaryRouteAnimation,
          curve: Curves.linearToEaseOut,
          // The curve must be flipped so that the reverse animation doesn't play
          // an ease-in curve, which iOS does not use.
          reverseCurve: Curves.linearToEaseOut.flipped,
-       ).drive(_kBottomUpTween),
+       )).drive(_kBottomUpTween);
        _secondaryPositionAnimation =
-           (linearTransition
-             ? secondaryRouteAnimation
-             : CurvedAnimation(
-                 parent: secondaryRouteAnimation,
+           (widget.linearTransition
+             ? widget.secondaryRouteAnimation
+             : _secondaryPositionCurve = CurvedAnimation(
+                 parent: widget.secondaryRouteAnimation,
```
Here you are never reassigning `null` to the curved animation. This means you might dispose them twice

ValentinVignal reviewed (commented):
I have the same comments as for https://github.com/flutter/flutter/pull/147133

Dimilkalathiya commented:
@ValentinVignal Again, thanks for review :) i'll add fix things as per your suggestions

Dimilkalathiya commented:
@polina-c Do we need test here? since it does not change outside behaviour, if so what kind of test would be good fit for this?

polina-c commented:
> @polina-c Do we need test here? since it does not change outside behaviour, if so what kind of test would be good fit for this?

Since this PR fixes a leak, it would be great to opt-in a test to track this fixed leak.

Dimilkalathiya committed:
changes as per suggestion and new testcase

Dimilkalathiya commented:
@polina-c I have added test-case as discussed, let me know we are good go.

At first i tried to add opt-in into existing testcase but it seems all `CupertinoPageRoute` test are all paired with `CupertinoApp` which it self leaks `CurvedAnimation` so i had to create new testcase with `MaterialApp` would that be okay?

Also, added null assigned to curve after disposal

polina-c commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -2431,6 +2432,52 @@ void main() {
       expect(tester.getBottomRight(find.byType(Placeholder)).dx, 390.0);
     });
   });
+  testWidgets(
```
Add empty line in front of test.

polina-c commented:
> @polina-c `framework_tests_impeller` seems to be failing and from looking at logs `editable_text_cursor_test.dart` has something to do with this, i ran tests but except of leaks logs test-cases are passing.
> 
> does this has something to do with current changes in PR?

It seems they passed now.

Dimilkalathiya committed:
add space before test-case

Dimilkalathiya committed:
revert flutter_test_config.dart

Dimilkalathiya committed:
remove space 🫣

polina-c commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -453,44 +453,97 @@ class CupertinoPageTransition extends StatelessWidget {
+
+class _CupertinoFullscreenDialogTransitionState extends State<CupertinoFullscreenDialogTransition> {
+
+    // When this page is coming in to cover another page.
```
Remove these two spaces and the entire line on top.
(hopefully we will have auto-formatting for flutter soon and will stop spending effort on this)

polina-c commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -453,44 +453,97 @@ class CupertinoPageTransition extends StatelessWidget {
+    if (oldWidget.primaryRouteAnimation != widget.primaryRouteAnimation
+    || oldWidget.secondaryRouteAnimation != widget.secondaryRouteAnimation
+    || oldWidget.child != widget.child || oldWidget.linearTransition != widget.linearTransition) {
+    _disposeCurve();
```
add indentation

polina-c commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -453,44 +453,97 @@ class CupertinoPageTransition extends StatelessWidget {
+  late Animation<Offset> _secondaryPositionAnimation;
+  // Curve of primary page which is coming in to cover another page.
+  CurvedAnimation? _primaryPositionCurve;
+  // Curve of secondary page which is becoming covered by another page.
```
// -> ///
for all comments for fields

ValentinVignal commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -453,44 +453,97 @@ class CupertinoPageTransition extends StatelessWidget {
+    }
+  }
+
+   @override
+  void dispose() {
+    super.dispose();
+    _disposeCurve();
+  }
```
The body of dispose should end with `super.dispose()`

ValentinVignal commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -2431,6 +2432,53 @@ void main() {
+  'Fullscreen route does not leak CurveAnimation', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Builder(
+          builder: (BuildContext context) {
```
I think you can remove this `Builder` widget, right?

Dimilkalathiya replied:
Actually we need this for accessing `context` for `Navigator.push`

Dimilkalathiya committed:
changes as per review

Dimilkalathiya commented:
@polina-c @ValentinVignal Added mentioned changes 👀

polina-c commented:
This is not addressed yet: https://github.com/flutter/flutter/pull/147168#discussion_r1580089692

Dimilkalathiya committed:
added indention

Dimilkalathiya commented:
@polina-c please check now

polina-c committed:
Merge branch 'master' into fix_full_dialg_transition_leak
